File: baek.17822.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static int[][] arr;
    static int n, m, t;
    static int[] dx = new int[]{-1,0,1,0};
    static int[] dy = new int[]{0,1,0,-1};
    static class Pair {
        int a, b;
        Pair(int a, int b){this.a = a; this.b = b;}
        int first(){return this.a;}
        int second(){return this.b;}
    }
    static void spin(int idx, int rot, int dir) {
        if(dir==1)
            rot *= -1;
        int[] tmp = arr[idx].clone();
        for(int i=0; i<arr[idx].length; i++){
            arr[idx][i] = i-rot>=0 ? tmp[(i-rot)%m] : tmp[m + (i-rot)%m];
        }
        return;
    }
    static boolean checka(int a){return a >= 1 && a <= n;}
    static int checkb(int b) {return b<0? m-1 : (b==m ? 0 : b);}
    static void bfs(int a, int b){
        Queue<Pair> q = new LinkedList<>();
        boolean[][] v = new boolean[n+1][m];
        for(int i=0; i<=n; i++)
            Arrays.fill(v[i], false);
        q.offer(new Pair(a, b));
        v[a][b] = true;
        while(!q.isEmpty()){
            Pair cur = q.poll();
            for(int i=0; i<4; i++){
                int da = cur.first() + dx[i];
                int db = checkb(cur.second() + dy[i]);
                if(checka(da) && arr[da][db]!=0 && arr[da][db]==arr[a][b] && !v[da][db]){
                    arr[da][db] = 0;
                    v[da][db] = true;
                    q.offer(new Pair(da, db));
                }
            }
        }
    }
    public static void main(String[] args) {
        int x, d, k, sum, cnt;
        double aver;
        n = sc.nextInt();
        m = sc.nextInt();
        t = sc.nextInt();
        arr = new int[n+1][m];
        for(int i=1; i<=n; i++)
            for(int j=0; j<m; j++)
                arr[i][j] = sc.nextInt();

        for(int tc=0; tc<t; tc++) {
            x = sc.nextInt();
            d = sc.nextInt();
            k = sc.nextInt();
            for (int i = 1; i <= n; i++)
                if (i % x == 0)
                    spin(i, k, d);
            boolean same = false;
            for(int a=1; a<=n; a++){
                for(int b=0; b<m; b++){
                    for(int c=0; c<4; c++){
                        int da = a + dx[c];
                        int db = checkb(b + dy[c]);
                        if(checka(da) && arr[a][b]!=0 && arr[da][db]!=0 && arr[da][db]==arr[a][b]){
                            bfs(a, b);
                            same = true;
                            arr[a][b] = 0;
                        }
                    }
                }
            }
            if(!same){
                sum = 0;
                cnt = 0;
                for(int a=1; a<=n; a++){
                    for(int b=0; b<m; b++){
                        if(arr[a][b]!=0){
                            sum += arr[a][b];
                            cnt++;
                        }
                    }
                }
                aver = (double)sum / (double)cnt;
                for(int a=1; a<=n; a++){
                    for(int b=0; b<m; b++) {
                        if(arr[a][b]!=0){
                            if((double)arr[a][b]>aver)
                                arr[a][b]--;
                            else if((double)arr[a][b]<aver)
                                arr[a][b]++;
                        }
                    }
                }
            }
        }
        sum = 0;
        for(int a=1; a<=n; a++)
            for(int b=0; b<m; b++)
                sum += arr[a][b];
        System.out.println(sum);
    }
}

File: baek_17822.cpp
#include <stdio.h>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

vector<vector<pair<int,int>>> v (1, vector<pair<int,int>> (1, make_pair(0,0)));
vector<vector<bool>> visited (1, vector<bool> (1, false));
int n, m, t, num, x, d, k, ans;
int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, -1, 0, 1};

bool checka(int a){
    if(a<1 || a>n)
        return false;
    else
        return true;
}
int checkb(int b){
    if(b==-1)
        return m-1;
    else if(b==m)
        return 0;
    else
        return b;
}


void spin(int dir, int idx, int cnt){
    int d =  dir==0 ? 1 : -1;
    cnt *= d;
    for(int i=0; i<m; i++){
        v[idx][i].first = (v[idx][i].first + cnt) % m;
        if(v[idx][i].first<0)
            v[idx][i].first = m + v[idx][i].first;
    }
}

void bfs(int a, int b){
    for(int i=1; i<=n; i++)
        for(int j=0; j<m; j++)
            visited[i][j] = false;
    queue<pair<int,int>> q;
    q.push(make_pair(a,b));
    visited[a][b] = true;
    while(!q.empty()){
        pair<int,int> cur = q.front();
        q.pop();
        for(int i=0; i<4; i++){
            int da = cur.first + dx[i];
            int db = checkb(cur.second + dy[i]);
            if(checka(da) && v[da][db].second==v[a][b].second && !visited[da][db]){
                q.push(make_pair(da,db));
                visited[da][db] = true;
                v[da][db].second = 0;
            }
        }
    }
    v[a][b].second = 0;
}

int main(void){
    ans = 0;
    scanf("%d %d %d", &n, &m, &t);
    for(int i=1; i<=n; i++){
        vector<pair<int,int>> vv;
        vector<bool> vvv;
        v.push_back(vv);
        visited.push_back(vvv);
        for(int j=0; j<m; j++){
            scanf("%d", &num);
            v[i].push_back(make_pair(j, num));
            visited[i].push_back(false);
        }
    }
    for(int i=0; i<t; i++){
        bool samenum = false;
        scanf("%d %d %d", &x, &d, &k);
        for(int j=1; j<=n; j++){
            if(j%x==0){
                spin(d, j, k);
                sort(v[j].begin(), v[j].end());
            }
        }
        for(int a=1; a<=n; a++){
            for(int b=0; b<m; b++){
                if(v[a][b].second!=0){
                    for(int c=0; c<4; c++){
                        int da = a + dx[c];
                        int db = checkb(b + dy[c]);
                        if(checka(da) && v[a][b].second!=0 && v[a][b].second==v[da][db].second){
                            bfs(a,b);
                            samenum = true;
                        }
                    }
                }
            }
        }
        if(!samenum){
            int total = 0;
            int cnt = 0;
            double aver = 0;
            for(int a=1; a<=n; a++){
                for(int b=0; b<m; b++){
                    if(v[a][b].second!=0){
                        total += v[a][b].second;
                        cnt++;
                    }
                }
            }
            if(cnt!=0){
                aver = (double)total / (double)cnt;
            }
            for(int a=1; a<=n; a++){
                for(int b=0; b<m; b++){
                    if(v[a][b].second!=0){
                        if(aver>(double)v[a][b].second)
                            v[a][b].second++;
                        else if(aver<(double)v[a][b].second)
                            v[a][b].second--;
                    }
                }
            }
        }
    }
    for(int i=1; i<=n; i++)
        for(int j=0; j<m; j++)
            ans += v[i][j].second;
    printf("%d\n", ans);
    return 0;
}